packages/pkl.lua/lua.pkl

```diff
@@ -294,3 +332,560 @@ class Renderer extends ValueRenderer {
+    }
+
+  function setKey(value, token: RegexMatch): ChildParseState =
+    (this) {
+      key = value
+      op = "key"
+      negate = null
+      mapStart =
+        // don't update mapStart if this will become a listing element
+        if (isListIndex(value)) super.mapStart
+        else super.mapStart ?? token
+    }
+}
+
+// endregion
+// region Tokens
+
+// Regex that matches a single Lua token, or an invalid character.
+// This regex assumes line endings have already been normalized, so no carriage returns exist.
+// Error states:
+// - Group 2 is ""
+// - Group 3 is not "\""?
+// - Group 4 is not "'"?
+// - Group 6 is ""
+// - Last group is non-null
+local const tokenRegex: Regex = Regex(##"""
+  (?x-uU)
+    --(?:               # comment
+      \[(?<eq1>=*)\[    # long comment (equals are GROUP 1)
+      (?>(?>
+        [^]]
+      | ](?!\k<eq1>])
+      )*)
+      (]\k<eq1>]|\z)    # match end or EOF (GROUP 2)
+    | .*                # short comment
+    )
+  | [\w&&\D]\w*         # identifier
+  | "(?>(?>             # short literal string (")
+      [^\\\n"]
+    | \\(?>z[\ \f\n\t\x0b]*|x\p{XDigit}{2}|\d{1,3}|u\{\p{XDigit}+}|(?s:.)|\z)
+    )*)("|(?s:.)|\z)    # match string truncated by newline or EOF (GROUP 3)
+  | '(?>(?>             # short literal string (')
+      [^\\\n']
+    | \\(?>z[\ \f\n\t\x0b]*|x\p{XDigit}{2}|\d{1,3}|u\{\p{XDigit}+}|(?s:.)|\z)
+    )*)('|(?s:.)|\z)    # match string truncated by newline or EOF (GROUP 4)
+  | \[(?<eq2>=*)\[      # long literal string (equals are GROUP 5)
+    (?>(?>
+      [^]]
+    | ](?!\k<eq2>])
+    )*)
+    (]\k<eq2>]|\z)      # match end or EOF (GROUP 6)
+    # for the numeric literals, they consume extra periods and hex digits and throw a parse error
+    # this regex aims to match everything Lua 5.3 itself will tokenize as numeric
+  | \.?0[xX](?:(?:[pP][-+]?)?[.\p{XDigit}])*   # hex numeric literal
+  | \.?\d(?:(?:[eE][-+]?)?[.\p{XDigit}])*   # dec numeric literal
+  | [-+*%^\#&|(){}\[\];,]   # single-char operators
+  | <[<=]? | >[>=]? | //? | ~=? | ==? | ::? | \.{1,3}   # multi-char operators
+  | ([^\ \f\n\t\x0b])   # invalid token (last group)
+  """##)
+
+// checks the error states documented on tokenRegex
+// returns the same token, or throws an error
+local const function validateToken(token: RegexMatch, source: String, uri: Uri?): RegexMatch =
+  if (token.groups[2]?.value == "") throwError("expected ]\(token.groups[1]!!.value)], found EOF", source, uri, token.groups[2]!!)
+  else let (g = token.groups[3]) if (g != null && g.value != "\"") throwExpected("\"", source, uri, g)
+  else let (g = token.groups[4]) if (g != null && g.value != "'") throwExpected("'", source, uri, g)
+  else if (token.groups[6]?.value == "") throwError("expected ]\(token.groups[5]!!.value)], found EOF", source, uri, token.groups[6]!!)
+  else let (g = token.groups.last) if (g != null) throwError("illegal token \(g.value)", source, uri, g)
+  else token
```

Comment:
For consistency sake: error messages should start with a capitalized word.

```suggestion
  if (token.groups[2]?.value == "") throwError("Expected ]\(token.groups[1]!!.value)], found EOF", source, uri, token.groups[2]!!)
  else let (g = token.groups[3]) if (g != null && g.value != "\"") throwExpected("\"", source, uri, g)
  else let (g = token.groups[4]) if (g != null && g.value != "'") throwExpected("'", source, uri, g)
  else if (token.groups[6]?.value == "") throwError("Expected ]\(token.groups[5]!!.value)], found EOF", source, uri, token.groups[6]!!)
  else let (g = token.groups.last) if (g != null) throwError("Illegal token \(g.value)", source, uri, g)
  else token
```

Reply:
Done